### ISSUE TITLE
Investigate Xcode project cherry-picking for caching

### ISSUE THREAD

**Issue by lacasseio** [nokeedev/gradle-native#661]
Let's investigate the Xcode project cherry-picking for caching, see https://github.com/nokeedev/gradle-native/issues/660.

**Comment by lacasseio** [comment]
Will this work as an out-of-box solution? There are some concerns with this approach regarding inter-project reaching which would not allow this feature to work. There are also some concerns about how to properly configure the build settings to allow the builds to split. The build settings override was a manual process. We need to create a deterministic process that works for all projects. We also need to correctly choose which files to cache as well as use the right inputs for the cache key.

We should spend a bit more time trying it out with other projects. At the very least, we could handle the Cocoapods project as a cacheable unit. A more complicated project such as a project with an intermediate project between Pods and the product project could make this process more complicated.

We would also have to see if circular dependencies between projects could happen, say target T1 in project A depends on target T2 in project B which depends on target T3 in project A. There is a circular dependency between projects but not between targets which should, in theory, be allowed.

**Comment by lacasseio** [comment]
It seems this solution is a bit what Carthage uses. Carthage will build upon resolving and then rely on file on disk instead of actual projects. Although it would mostly avoid costly rebuild, the problem is usually centered around branch switching which can cause Xcode to trash the build (or cause users to forget to update the dependencies causing build failures).

**Comment by lacasseio** [comment]
We continued the investigation and came up with the following process for now:

If a workspace is present, we map each Xcode project to a Gradle project. The Gradle root project will stand as the Xcode workspace and will delegate to `xcodebuild` using the `-workspace` flag. There isn't much we can do so it will be the escape hatch until we can validate the other mapping.

For each project, we map each Xcode target to a task. Those tasks will rely on Gradle dependency engine to pull derived data from other projects (external dependencies). The tasks will track the inputs and cache the outputs for a local build plan. The local build plan refers to the local target and its local dependencies. Any external dependencies are pulled via the dependency engine. This means that target A which depends on a local target B will cache A and B when building A and only B when building B. In theory, the target B will be cached twice but in practice it should not matter much.

For each project, we map the Xcode scheme to a task for each action (e.g. build, run/launch, test, profile, archive, analyze). Those tasks will look a lot like the target task. We can't really reuse the output from those target tasks because we delegate to xcodebuild which doesn't allow us to slice and dice the build plan. In theory we will over cache but in practice it shouldn't matter much.

Xcodebuild sometimes shows implicit schemes (schemes that do not exist but would normally be autogenerated). Although we have a good idea on how those are implied, we will only make basic assumptions until we have more vision in this solution.

With this approach, we should be able to cache part of the build without any effort from the users. Note that we will focus on local caching (when projects are in the exact location) because remote cache requires a bit more manipulation which can be improved later.

**Comment by lacasseio** [comment]
Not an easy task to accomplish mostly because of all the additional PBX types I need to support. I focused on a very small application, e.g. [HackerNews](https://github.com/amitburst/HackerNews), which contains 2 projects, e.g. main and Pods. The focus was put on detecting the input and outputs of a target cluster. In this case, we use the `Pods-HackerNews` target which depends on the `Firebase` target (inside the same `Pods` project). It takes [~2s to build](https://scans.gradle.com/s/fipxcxevdsh7y/performance/build#build-time-total). When we use the up-to-date detection from Gradle, we can avoid rebuilding which takes [~0.6s to build/check](https://scans.gradle.com/s/codsffgn5wedw/performance/build#build-time-total). Gradle's up-to-date check basically avoids the execution of the following xcodebuild command:

```
xcodebuild -project `pwd`/Pods/Pods.xcodeproj -target Pods-HackerNews PODS_BUILD_DIR=`pwd`/build/subprojects/Pods-3vq31kw6caact/tmp/Pods-HackerNews/derivedData/Build/Products BUILD_DIR=`pwd`/build/subprojects/Pods-3vq31kw6caact/tmp/Pods-HackerNews/derivedData/Build/Products BUILD_ROOT=`pwd`/build/subprojects/Pods-3vq31kw6caact/tmp/Pods-HackerNews/derivedData/Build/Products PROJECT_TEMP_DIR=`pwd`/build/subprojects/Pods-3vq31kw6caact/tmp/Pods-HackerNews/derivedData/Build/Intermediates.noindex/Pods.build OBJROOT=`pwd`/build/subprojects/Pods-3vq31kw6caact/tmp/Pods-HackerNews/derivedData/Build/Intermediates.noindex SYMROOT=`pwd`/build/subprojects/Pods-3vq31kw6caact/tmp/Pods-HackerNews/derivedData/Build/Products -sdk iphoneos -configuration Debug CODE_SIGN_IDENTITY="" CODE_SIGNING_REQUIRED=NO CODE_SIGN_ENTITLEMENTS="" CODE_SIGNING_ALLOWED="NO"
```

The xcodebuild command seems to take around ~1.2s for a fully up-to-date invocation. This means Gradle avoids around 0.6s of work (200% improvement).

The interesting part is not the up-to-date scenario but when the mtime changes without any file content changes. Suppose the following scenario:
```
$ rm -rf Pods
$ pod install
$ <rebuild>
```

For the same xcodebuild command that was previously built (before reinstalling Cocoapods), Gradle still takes [~0.6s](https://scans.gradle.com/s/lluygb4l2nic2/performance/build#build-time-total) as no input or output files changed. However, xcodebuild takes ~1.7s as it performs a rebuild since the mtime of the Pods file changed. In this case, Gradle avoids 1.1s of work (283% improvement).

Deleting the `build` directory would also give the advantage to Gradle when the build cache is supported (at the very least we could support the local build cache now) as we would avoid invoking `xcodebuild` and restore the binaries from the cache.

Finally, it's important to note the improvements exposed here are preliminary and the project used is quite trivial. The VLC project or more complex would better represent the actual improvement. We will now focus on finishing this code so we can perform real-life tests. We are also missing the implicit dependencies which would tie the Gradle project together. We will work on this a bit later. The current focus will be on properly detecting the input/output of a target cluster local to the project.

**Comment by lacasseio** [comment]
A quick test on [vlc-ios project](https://github.com/videolan/vlc-ios) shows the 2m 19s build time for target `Pods-VLC-iOS` and a 3s up-to-date check on the subsequent build. This test also made me think that we will need to add the files (or at least the serialized configuration) that enable the build (xconfig, pbxproj, etc). I mentioned the serialized configuration as it would be possible to serialize the values in the files to shield against an out-of-order scenario where the configuration file changes but the effective data are still the same. Given we already load pretty much everything in memory, we could _simply_ make sure the model is good enough for input snapshot. We may use another model as well for input snapshotting.

As mentioned the downside of the target clustering is the duplication of build tasks. For instance, building `MobileVLCKit` target will not reuse the work already done under `Pods-VLC-iOS`. However, we assume that in most cases, target cluster from a single project are often reused through out the build so the impact should be fairly minimal. Also, there is a good chance we could build both `Pods-VLC-iOS` and `MobileVLCKit` in parallel (using the Worker API) since we already isolate the derived data per target.

**Comment by lacasseio** [comment]
With PR https://github.com/nokeedev/gradle-native/pull/673, we can now build full projects with Gradle. There are still some issues with the correctness which we highlighted in the other issues. We ran some early experimentation and we see the following:

**HackerNews**: 
- Clean build (xcodebuild vs Gradle/Nokee): 11.1874s vs 6s
- Up-to-date build (xcodebuild vs Gradle/Nokee): 1.274s vs 0.5374s

**VLC**:
- Clean build (xcodebuild vs Gradle/Nokee): 3m37s vs 3m22s
- Up-to-date build (xcodebuild vs Gradle/Nokee): 10.977s vs 2s

**Expenso**
- Clean build (xcodebuild vs Gradle/Nokee): 1m59s vs 1m47s
- Up-to-date build (xcodebuild vs Gradle/Nokee): 1.76s vs 0.580s

Unfortunately, we don't support cross-project references meaning we can't build projects like **newsyc**. Although we lack some correctness in the build, the additional work won't have a significant impact on the configuration performance. Either way, we would still recommend users to use the configuration cache which should remove any configuration delays.
